core/src/main/scala/caliban/execution/Field.scala

```diff
@@ -121,68 +111,72 @@ object Field {
               None
             )
           }
-        case FragmentSpread(name, directives)                           =>
-          val resolvedDirectives = directives.map(directive =>
-            directive.copy(arguments = resolveVariables(directive.arguments, variableDefinitions, variableValues))
-          )
-
-          if (checkDirectives(resolvedDirectives)) {
-            fragments
-              .get(name)
-              .foreach { f =>
-                val t = innerType.possibleTypes
-                  .flatMap(_.find(_.name.contains(f.typeCondition.name)))
-                  .orElse(rootType.types.get(f.typeCondition.name))
-                  .getOrElse(fieldType)
-                if (f.selectionSet.nonEmpty) {
-                  loop(f.selectionSet, t).fields.foreach { field =>
-                    val _field =
-                      if (field._condition.isDefined) field
+        case fr @ FragmentSpread(name, directives)                      =>
+          val fields = memoizedFragments.getOrElseUpdate(
+            fr, {
+              val resolvedDirectives = directives.map(resolveDirectiveVariables(variableValues, variableDefinitions))
+
+              if (checkDirectives(resolvedDirectives)) {
+                fragments
+                  .get(name)
+                  .map { f =>
+                    val t = innerType.possibleTypes
+                      .flatMap(_.find(_.name.contains(f.typeCondition.name)))
```

Comment:
This is the part that concerns me the most, since `innerType` will be different each time we are here. If we memoize this, it seems that it won't take the new `innerType` into consideration. But in all cases that I tested this against, `innerType.possibleTypes` was `None`, so I'm not sure what needs to be done for it to be `Some(...)`